src/coreclr/jit/lowerxarch.cpp

```diff
@@ -8638,6 +8649,78 @@ void Lowering::ContainCheckCast(GenTreeCast* node)
 #endif // !defined(TARGET_64BIT)
 }
 
+//------------------------------------------------------------------------
+// TryLowerAndOrToCCMP : Lower AND/OR of two conditions into test + CCMP + SETCC nodes.
+//
+// Arguments:
+//    tree - pointer to the node
+//    next - [out] Next node to lower if this function returns true
+//
+// Return Value:
+//    false if no changes were made
+//
+bool Lowering::TryLowerAndOrToCCMP(GenTreeOp* tree, GenTree** next)
+{
+    assert(tree->OperIs(GT_OR, GT_AND));
+
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return false;
+    }
+
+    GenTree* lastNode = BlockRange().LastNode();
+
+    if (!lastNode->OperIs(GT_JTRUE, GT_RETURN))
+    {
+        return false;
+    }
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    if (!op1->OperIs(GT_NE, GT_EQ) || !op2->OperIs(GT_NE, GT_EQ) ||
+        (lastNode->OperIs(GT_JTRUE) && ((tree->OperIs(GT_OR) && (!op1->OperIs(GT_NE) || !op2->OperIs(GT_NE))) ||
+                                        (tree->OperIs(GT_AND) && (!op1->OperIs(GT_EQ) || !op2->OperIs(GT_EQ))))) ||
+        (lastNode->OperIs(GT_RETURN) && ((tree->OperIs(GT_OR) && (!op1->OperIs(GT_EQ) || !op2->OperIs(GT_EQ))) ||
+                                         (tree->OperIs(GT_AND) && (!op1->OperIs(GT_NE) || !op2->OperIs(GT_NE))))))
+    {
+        return false;
+    }
```

Comment:
Here too the `JTRUE/RETURN` checks do not make sense to me. Imagine having
```
var foo = (x == 0) & (y == 0).
var bar = (x != 0) | (y != 0);
```
in any basic block. The transformation should be able to kick in for this pattern regardless of what type of basic block it appears in.